### DISCUSSION TITLE
how  Crazyflie 2.1 or Crazyflie Bolt 1.1 printf lighthouse's position info, x, y, z

### DISCUSSION THREAD

**Discussion by justinleeyang** [bitcraze/discussions#1699] [General]
Hi, 

how  Crazyflie 2.1 or Crazyflie Bolt 1.1 printf lighthouse's position info, x, y, z

**Comment by ToveRumar** [comment]
Hi @justinleeyang!

Please refer to this documentation about how to set up a logging configuration when using the lib; https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/user-guides/sbs_connect_log_param/

If you want your log to be visible when using the python client you can use this user guide to set it up https://www.bitcraze.io/documentation/repository/crazyflie-clients-python/master/userguides/userguide_client/logblocks_tab/

I hope this helps you set up the logging you want for your system!

**Reply by justinleeyang** [reply]
I want to the f405 output the position information, No CF client python .

**Reply by ToveRumar** [reply]
Hi @justinleeyang !

Ok, Im not sure I understand your use case. 
To where do you need to print? Have you attached something to a log port of the STM32 and want to get the log output from there?  

If you just need to get the logs (printouts)  programatically from the Crazyflie (STM32), then  logging with the using the lib is the most straight forward solution. 
Just connect your Crazyflie with USB or Crazyradio and follow the instructions in; https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/user-guides/sbs_connect_log_param/
The params you want to print are already added to the logging framework so you do not need to change the Crazyflie Firmware.


If this is not what you are looking to do, please explain your use case in a bit more detail so I can help you reach a solution.

**Reply by justinleeyang** [reply]
I have a robotic car with MCU based on Freertos. It communicate with uart. Don't install python cf lib. So need crazyflie output xyz with uart to my car mcu

**Reply by ToveRumar** [reply]
Thank you for explaining your use case!

In this case I would use either CPX over uart (https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/cpx/) or do it like you described by sending only the positions over uart. There is an example app for doing this here ( https://github.com/knmcguire/crazyflie-firmware/tree/bolt-pos-through-uart/examples/app_uart). There Is also a description on how to set up uart communication to the STM32 here https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/development/uart_communication/. I think this documentation can help you put together what you  need to achieve your goal!

**Reply by justinleeyang** [reply]
Thanks, and how to output the position data over stm32's E_TX2 adn E_RX2, Please indicate the detailed steps.

![Uploading image.png…]()

Is this example over E_TX2 adn E_RX2 ? 
https://github.com/knmcguire/crazyflie-firmware/tree/bolt-pos-through-uart/examples/app_uart

![Uploading image.png…]()

**Reply by justinleeyang** [reply]
Thanks, and how to output the position data over stm32's E_TX2 adn E_RX2, Please indicate the detailed steps.

![image](https://github.com/user-attachments/assets/febf465a-f71d-43d6-b824-fd8a66ae66b2)


Is this example over E_TX2 adn E_RX2 ? 

> #define UART_TX_PIN 28 // these might need to be changed
> #define UART_RX_PIN 29 // these might need to be changed

https://github.com/knmcguire/crazyflie-firmware/tree/bolt-pos-through-uart/examples/app_uart

![image](https://github.com/user-attachments/assets/fd287728-27d8-4227-894d-922099ac8a3a)

**Reply by ToveRumar** [reply]
Hi @justinleeyang !
No this example uses another set of pins. 
If you want to use e_tx2 and e_rx2 they are mapped to PA2 and PA3 on the STM. 

These pins are the UART2 pins already so I think you could re-use the setup for UART2 in the AI-deck driver (https://github.com/knmcguire/crazyflie-firmware/blob/bolt-pos-through-uart/src/deck/drivers/src/aideck.c).

While looking around I also found this driver for setting up UART2 for CPX (https://github.com/knmcguire/crazyflie-firmware/blob/bolt-pos-through-uart/src/deck/drivers/src/cpx-host-on-uart2.c). Note that this driver needs to be force included at build time using the menuconfig.
